sled-agent/src/rack_setup/config.rs

```diff
@@ -0,0 +1,52 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Interfaces for working with RSS config.
+
+use crate::config::ConfigError;
+use crate::params::{DatasetEnsureBody, ServiceRequest};
+use serde::Deserialize;
+use serde::Serialize;
+use std::net::SocketAddr;
+use std::path::Path;
+
+/// Configuration for the "rack setup service", which is controlled during
+/// bootstrap.
+///
+/// The Rack Setup Service should be responsible for one-time setup actions,
+/// such as CockroachDB placement and initialization.  Without operator
+/// intervention, however, these actions need a way to be automated in our
+/// deployment.
+///
+/// By injecting this (optional) configuration into the bootstrap agent, it
+/// can act as a stand-in initialization service.
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+pub struct SetupServiceConfig {
+    #[serde(default, rename = "request")]
+    pub requests: Vec<SledRequest>,
+}
+
+/// A request to initialize a sled.
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+pub struct SledRequest {
+    /// The Sled Agent address receiving these requests.
+    pub sled_address: SocketAddr,
+
+    /// Partitions to be created.
+    #[serde(default, rename = "partition")]
+    pub partitions: Vec<DatasetEnsureBody>,
```

Comment:
Not really related to this PR, but why are these called "partitions"? My understanding is that these define a combination of a zpool, a kind of dataset to put on top (usually a filesystem), and an address for a service inside the zone to listen on. That really sounds like a "zoned service" or something, so I'm just trying to get clear on the terminology here.

Reply:
Honestly, it was just poor naming on my part. I viewed them as "splitting up the logical space on a disk - as in, we're partitioning it".

... However, this is was a poor naming decision, as a "partition" is a distinct thing from a "dataset".

I'll submit a follow-up PR to rename "partition" to "dataset" across the whole repo.

Reply:
https://github.com/oxidecomputer/omicron/issues/878